sw/device/lib/crypto/impl/hash.c

```diff
@@ -6,13 +6,57 @@
 
 #include <stdbool.h>
 
+#include "sw/device/lib/base/hardened_memory.h"
 #include "sw/device/lib/crypto/drivers/hmac.h"
 #include "sw/device/lib/crypto/drivers/kmac.h"
+#include "sw/device/lib/crypto/impl/sha2/sha512.h"
 #include "sw/device/lib/crypto/impl/status.h"
 
 // Module ID for status codes.
 #define MODULE_ID MAKE_MODULE_ID('h', 'a', 's')
 
+/**
+ * Ensure that the hash context is large enough for the SHA-512 state struct.
+ */
+static_assert(sizeof(hash_context_t) >= sizeof(sha512_state_t),
+              "hash_context_t must be big enough to hold sha512_state_t");
+/**
+ * Ensure that the SHA-512 state struct is suitable for `hardened_memcpy()`.
+ */
+static_assert(sizeof(sha512_state_t) % sizeof(uint32_t) == 0,
+              "Size of sha512_state_t must be a multiple of the word size for "
+              "`hardened_memcpy()`");
+
+/**
+ * Save a SHA-512 state to a generic hash context.
+ *
+ * @param[out] ctx Generic hash context to copy to.
+ * @param state SHA-512 context object.
+ */
+static void sha512_state_save(hash_context_t *const OT_RESTRICT ctx,
```

Comment:
In other places we use the keyword directly in C. e.g.

```c
hash_context_t *const restrict ctx
```

Should this be as follows to match the expected `hardened_memcpy` parameters? Assuming the docstring doesn't need to be updated.

```c
sha512_state_save(hash_context_t *restrict ctx, const sha512_state_t *restrict state)
```